### PR TITLE
Security enhancements

### PR DESCRIPTION
**Pull request by petermcd**
Improved security

## Summary by Sourcery

Strengthen authentication, webhook handling, HTTP communication, and credential storage security across the library.

New Features:
- Validate authentication redirect URLs and reject unsafe or invalid values.
- Add HTTPS-only validation for webhook URLs before creation.
- Expose an is_authenticated property to check for a valid, non-expired access token.

Bug Fixes:
- Map additional HTTP status codes and network errors to appropriate Monzo-specific exceptions.
- Add tests covering authentication edge cases, webhook URL validation, and HTTP error handling.

Enhancements:
- Tighten access token refresh behaviour to proactively refresh tokens nearing expiry.
- Improve logging around authentication, token exchange, token refresh, and logout operations.
- Ensure filesystem credential storage uses restrictive file permissions and explicit typing.
- Refine attachment creation logic to validate supported file types more robustly and handle local file uploads correctly.

Build:
- Add explicit setuptools-based build-system configuration in pyproject and bump project version to 1.3.0.

Tests:
- Expand test coverage for authentication flows, webhook URL validation, and HTTP/URL error handling in the HttpIO layer.

### CODE REVIEW

File: monzo/authentication.py [hunk 1]
```diff
@@ -117,7 +125,7 @@ def make_request(
         Raises:
             MonzoHTTPError: On using an invalid method
         """
-        if self._access_token and self._access_token_expiry - time() < 0:
+        if self._access_token and self._access_token_expiry - time() < 60:
             self.refresh_access()
         if data is None:
             data = {}
```
[the remark this answers is not in the document]
Comment (petermcd) [reply monzo/authentication.py:131]: fixed in next commit

File: monzo/authentication.py [hunk 2]
```diff
@@ -264,7 +282,7 @@ def _populate_tokens(self, response: REQUEST_RESPONSE_TYPE) -> None:
         Args:
             response: Response from an auth request.
         """
-        logger.debug(msg="Populating tokens")
+        logger.info(msg="Populating tokens")
         self._access_token = response["data"]["access_token"]
         self.access_token_expiry = response["data"]["expires_in"]
         self._refresh_token = ""
```
[the remark this answers is not in the document]
Comment (petermcd) [reply monzo/authentication.py:290]: this is populated in the next line

File: monzo/handlers/filesystem.py [hunk 3]
```diff
@@ -52,12 +53,14 @@ def store(
...
             "client_secret": client_secret,
             "expiry": expiry,
             "refresh_token": refresh_token,
         }
-        with open(self._file, "w") as handler:
-            handler.write(dumps(content))
+        fd: int = os.open(path=self._file, flags=os.O_WRONLY | os.O_CREAT | os.O_TRUNC, mode=0o600)
+        with os.fdopen(fd, mode="w") as handler:
+            handler.write(dumps(obj=content))
+        os.chmod(path=self._file, mode=0o600)
```
[the remark this answers is not in the document]
Comment (petermcd) [reply monzo/handlers/filesystem.py:66]: this is incorrect

File: pyproject.toml [hunk 4]
```diff
@@ -27,6 +27,10 @@ repository = "https://github.com/petermcd/monzo-api.git"
 issues = "https://github.com/petermcd/monzo-api/issues"
 documentation = "https://monzo-api.readthedocs.io"
 
+[build-system]
+requires = ["setuptools>=42"]
+build-backend = "setuptools.build_meta"
```
[the remark this answers is not in the document]
Comment (petermcd) [reply pyproject.toml:32]: updated in next commit

File: monzo/authentication.py [hunk 5]
```diff
@@ -224,12 +242,12 @@ def state_token(self) -> str:
...
-            with open(tmp_file_path, "w") as fh:
-                state_token = secrets.token_urlsafe(64)
+            with open(file=tmp_file_path, mode="w") as fh:
+                state_token: str = secrets.token_urlsafe(64)
                 fh.write(state_token)
                 fh.flush()
-        with open(tmp_file_path, "r") as fh:
-            state_token = fh.read()
+        with open(file=tmp_file_path, mode="r") as fh:
+            state_token: str = fh.read()
```
[the remark this answers is not in the document]
Comment (petermcd) [reply monzo/authentication.py]: permissions fixed in next commit

File: monzo/endpoints/attachment.py [hunk 6]
```diff
@@ -140,11 +140,12 @@ def create_attachment(cls, auth: Authentication, transaction_id: str, url: str)
...
-        if file_extension not in SUPPORTED_ATTACHMENT_EXTENSIONS:
+        ext: str = file_extension.lstrip(".")
+        if ext not in SUPPORTED_ATTACHMENT_EXTENSIONS:
             raise MonzoGeneralError("Unsupported file type")
-        file_type = SUPPORTED_ATTACHMENT_EXTENSIONS[file_extension]
-        if file_url.netloc:
-            file_type = Attachment._upload_file(auth=auth, url=url, file_type=file_type)
+        file_type: str = SUPPORTED_ATTACHMENT_EXTENSIONS[ext]
+        if not file_url.netloc:
+            file_type: str = Attachment._upload_file(auth=auth, url=url, file_type=file_type)
```
[the remark this answers is not in the document]
Comment (petermcd) [reply monzo/endpoints/attachment.py:148]: functionality not fully implemented
